[InstCombine] Preserve the nsw/nuw flags for (X | Op01C) + Op1C --> X + (Op01C + Op1C)

csstormq opened the pull request:
This patch simplifies `sdiv` to `udiv` by preserving the `nsw` flag for `(X | Op01C) + Op1C --> X + (Op01C + Op1C)` if the sum of `Op01C` and `Op1C` will not overflow, and preserves the `nuw` flag unconditionally.

Alive2 Proofs (provided by @nikic): https://alive2.llvm.org/ce/z/nrdCZT, https://alive2.llvm.org/ce/z/YnJHnH

csstormq committed:
[InstCombine] Preserve the nsw flag for (X | Op01C) + Op1C --> X + (Op01C + Op1C)

nikic reviewed (changes requested):
Please always provide alive2 proofs for InstCombine changes, see https://llvm.org/docs/InstCombineContributorGuide.html#proofs.

The transform as implemented is incorrect: https://alive2.llvm.org/ce/z/pEk7wC

nikic commented:
This would be a correct variant of the transform: https://alive2.llvm.org/ce/z/nrdCZT

csstormq commented:
> Please always provide alive2 proofs for InstCombine changes, see https://llvm.org/docs/InstCombineContributorGuide.html#proofs.
> 
> The transform as implemented is incorrect: https://alive2.llvm.org/ce/z/pEk7wC

Alive2 Proof: https://alive2.llvm.org/ce/z/CRKtkU

csstormq committed:
update test case using the update_test_checks.py script

nikic reviewed (changes requested):
Your proof is not for the transform you actually implement. The transform is still incorrect in general, you merely picked out one example where it is correct. I have already shared what you need to do to make it correct in my second comment. The necessary pre-condition can be implemented using the willNotOverflowSignedAdd helper.

csstormq committed:
fixup! Preserve the nsw flag iff the sum of Op01C and Op1C will not overflow

csstormq committed:
Add more test cases

nikic commented on llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp:
```diff
@@ -905,8 +905,17 @@ Instruction *InstCombinerImpl::foldAddWithConstant(BinaryOperator &Add) {
+    // For some cases, sdiv can be converted to udiv when the newly created add
+    // carrying the nsw flag is one of its operands.
+    if (willNotOverflowSignedAdd(Op01C, Op1C, Add))
+      NewAdd->setHasNoSignedWrap(Add.hasNoSignedWrap());
```
In the interest of completeness, can you please also preserve `nuw`? nuw can be preserved unconditionally, without an extra overflow check (see https://alive2.llvm.org/ce/z/YnJHnH).

nikic commented on llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp:
```diff
@@ -905,8 +905,17 @@ Instruction *InstCombinerImpl::foldAddWithConstant(BinaryOperator &Add) {
+    // Preserve the nsw flag iff the sum of Op01C and Op1C will not overflow
+    // so that there is a chance to make some other transformations.
+    // For some cases, sdiv can be converted to udiv when the newly created add
+    // carrying the nsw flag is one of its operands.
```
Remove this comment. You don't need to justify preserving nowrap flags.

nikic commented on llvm/test/Transforms/InstCombine/add.ll:
```diff
@@ -4137,5 +4137,25 @@ define i32 @fold_zext_nneg_add_const_fail2(i8 %x) {
+  %or = or disjoint i8 %x, -128
+  %add = add nsw i8 %or, -8
+  ret i8 %add
+}
```
Can you please find where the existing tests for this transform are (e.g. by commenting it out and seeing what fails), and move these tests there, together with the sdiv_to_udiv test you added below)?

csstormq replied:
After commenting it out, failed tests as shown below:
```
llvm/test/Transforms/InstCombine/add.ll:
  add_like_or_t0
  add_like_or_t2_extrause
  add_like_or_disjoint

llvm/test/Transforms/InstCombine/freeze.ll:
  early_freeze_test3

llvm/test/Transforms/InstCombine/sadd-with-overflow.ll:
  fold_with_distjoin_or

llvm/test/Transforms/LoopVectorize/interleaved-accesses.ll:
  PR27626_5

llvm/test/Transforms/PhaseOrdering/AArch64/matrix-extract-insert.ll:
  matrix_extract_insert_loop
```

Additionally, related commit: https://github.com/llvm/llvm-project/commit/a36bb7fd761942577b0d20d7f9b2bac57ffcd986

I think the surroundings of function add_like_or_* are good places to place these tests.

csstormq committed:
Resolve conversations

1. Preserve the nuw flag also

2. Remove unnecessary comment

3. Move test cases to the better place

csstormq committed:
Remove added new lines

nikic reviewed (approved):
LGTM

csstormq committed:
Resolve conversations

1. Save the overflow check if not needed

2. Add preserve nuw test case

nikic commented on llvm/test/Transforms/InstCombine/add.ll:
```diff
@@ -1510,6 +1510,46 @@ define i8 @add_like_or_t2_extrause(i8 %x) {
+; CHECK-NEXT:    ret i8 [[ADD]]
+;
+  %or = or disjoint i8 %x, 128
+  %add = add nuw i8 %or, 129
```
```suggestion
  %add = add nuw i8 %or, 127
```
The test as written is not great because it always returns poison. It's better to use a test case that can return non-poison results.

csstormq committed:
Fix test to not always return poison

csstormq commented:
@nikic Thank you very much for your review. I have learned a lot from your comments.
